src/service/url-replacements-impl.js

```diff
@@ -785,9 +785,10 @@ export class UrlReplacements {
       'CLIENT_ID': true,
       'QUERY_PARAM': true,
     };
+    const additionalURLParameters = element.getAttribute('data-amp-addparams');
     const whitelist = this.getWhitelistForElement_(
         element, supportedReplacements);
-    if (!whitelist) {
+    if (!whitelist && !additionalURLParameters) {
```

Comment:
So the paths below aren't necessary (or not always necessary) if we only have `aditionalURLParameters`.

- We don't need to build the href again if we've already done it
- We don't need a secure URL (no privacy issues)
- We don't need to expand the URL

We really need to move the `addParamsToUrl` stuff up, then we can do a few things:

```js
if (!whitelist && !additional) {
  return;
}
let herf = assertString(element[original_href_property] || getAttribute('href'));
if (!element[original_href_property]) {
  element[original_href_property) = href;
}
if (additional) {
  href = addParams(additional);
}
if (whitelist) {
  assertSecure();
  assertAllowedOrigin();
  href = expandSync();
}
element.href = href;